### ISSUE TITLE
Doris will scan all partitions when BinaryPredicate use default compatible type

### ISSUE THREAD

**Issue by spaces-X** [apache/incubator-doris#3887]
Create table:
```
create table t2 (dt int, id int, pv int sum) aggregate key(`dt`, `id`)
partition by range (dt) (
  partition p20200610 values less than ("20200610"),
  partition p20200611 values less than ("20200611"))
distributed by hash(`id`) buckets 1
properties("replication_num"="1");
```
Load Data:
```
insert into t2 (dt, id, pv) values (20200609, 1, 10), (20200609, 2, 20);
insert into t2 (dt, id, pv) values (20200610, 1, 10), (20200610, 2, 20);
```
Query1:
```
explain select * from t2 where dt=20200609;
|  0:OlapScanNode                                                   |
|   TABLE: t2                                                    |
|   PREAGGREGATION: OFF. Reason: No AggregateInfo                                  |
|   PREDICATES: `dt` = 20200609                                           |
|   partitions=1/2                only one partition
```
Query2:
```
explain select * from t2 where dt='20200609';

|  0:OlapScanNode                                                   |
|   TABLE: t2                                                    |
|   PREAGGREGATION: OFF. Reason: No AggregateInfo                                  |
|   PREDICATES: `dt` = 2.0200609E7                                         |
|   partitions=2/2                scan all partitions 
```
In query2, the compatible type of `whereclause`(`BinaryPredicate`) will be `Type.DOUBLE` and the `SlotRef` of 'dt' will be casted to CastExpr with double type, which cause the `canHashPartition` function return false. 
In this situation, we couldn't create `createPartitionFilter` successfully and scan all partitions.

CastExpr#canHashPartition
```
    public boolean canHashPartition() {
        if (type.isFixedPointType() && getChild(0).getType().isFixedPointType()) {
            return true;
        }
        if (type.isDateType() && getChild(0).getType().isDateType()) {
            return true;
        }
        return false;
    }
```

BinaryPredicate#getSlotBinding
```
public Expr getSlotBinding(SlotId id) {
        SlotRef slotRef = null;
        // check left operand
        if (getChild(0) instanceof SlotRef) {
            slotRef = (SlotRef) getChild(0);
        } else if (getChild(0) instanceof CastExpr && getChild(0).getChild(0) instanceof SlotRef) {
            if (((CastExpr) getChild(0)).canHashPartition()) {
                slotRef = (SlotRef) getChild(0).getChild(0);
            }
        }
     ...
    return null
}
```
Any advices for improving this situation?

By the way, why should we use the `canHashPartition` condition in `getSlotBinding`?
Could we safely remove `canHashPartition` in `getSlotBinding`?

**Comment by chaoyli** [comment]
If '20200609' should cast to int will be right.

**Comment by spaces-X** [comment]
> If '20200609' should cast to int will be right.

Yeah
If we change the logic in BinaryPredicate#getCmpType to add a rule like `(t1 = VARCHAR, t2 = FixedPointType: INT/BIGINT/...) -> (compatible type= FixedPointType: INT/BIGINT/...)`, will it be ok?

**Comment by chaoyli** [comment]
I am not very be skilled of it.  @EmmyMiao87 is a better object to answer the question.

**Comment by WingsGo** [comment]
> > If '20200609' should cast to int will be right.
> 
> Yeah
> If we change the logic in BinaryPredicate#getCmpType to add a rule like `(t1 = VARCHAR, t2 = FixedPointType: INT/BIGINT/...) -> (compatible type= FixedPointType: INT/BIGINT/...)`, will it be ok?

We can not simple change like this, you can see this PR https://github.com/apache/incubator-doris/pull/3527 , change the behavior will cause some problem, such as:select '1.1' = 1; if we cast it to int , it will return true.select '0.0' = 0; will throw a error

**Comment by spaces-X** [comment]
> > > If '20200609' should cast to int will be right.
> > 
> > 
> > Yeah
> > If we change the logic in BinaryPredicate#getCmpType to add a rule like `(t1 = VARCHAR, t2 = FixedPointType: INT/BIGINT/...) -> (compatible type= FixedPointType: INT/BIGINT/...)`, will it be ok?
> 
> We can not simple change like this, you can see this PR #3527 , change the behavior will cause some problem, such as:select '1.1' = 1; if we cast it to int , it will return true.select '0.0' = 0; will throw a error

I see.Thanks a lot.
So in this pr  [#3527](https://github.com/apache/incubator-doris/pull/3527) , String will cast to NumericType in Binary Predicate, which will cause the problem you mentioned : `select '1.1' = 1` and this pr can not be merged. 
If we want to avoid scaning all partitions, we need some futher more works to do on the partition type itself.
Right?

**Comment by WingsGo** [comment]
> > > > If '20200609' should cast to int will be right.
> > > 
> > > 
> > > Yeah
> > > If we change the logic in BinaryPredicate#getCmpType to add a rule like `(t1 = VARCHAR, t2 = FixedPointType: INT/BIGINT/...) -> (compatible type= FixedPointType: INT/BIGINT/...)`, will it be ok?
> > 
> > 
> > We can not simple change like this, you can see this PR #3527 , change the behavior will cause some problem, such as:select '1.1' = 1; if we cast it to int , it will return true.select '0.0' = 0; will throw a error
> 
> I see.Thanks a lot.
> So in this pr [#3527](https://github.com/apache/incubator-doris/pull/3527) , String will cast to NumericType in Binary Predicate, which will cause the problem you mentioned : `select '1.1' = 1` and this pr can not be merged.
> If we want to avoid scaning all partitions, we need some futher more works to do on the partition type itself.
> Right?

Yes, the PR didn't solve these problems, so if we need to avoid such situation, we much need much more test to ensure more situation is taken into conside.Currently we can only avoid by our sql.
